---
title: ''
author: ''
date: ''
output:
  pdf_document:
    keep_tex: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lastpage}
- \usepackage{lineno}
- \linenumbers
fontsize: 12pt
---

<!--- Define Headers and Footers --->
\fancypagestyle{plain}{%
  \renewcommand{\headrulewidth}{0pt}%
  \fancyhf{}%
  \fancyfoot[R]{\footnotesize Page \thepage\, of\, \pageref*{LastPage}}
  \setlength\footskip{0pt}
}
\pagestyle{plain}


```{r Fig1, warning=FALSE,echo=F, fig.width=7,fig.height=6, message=F, fig.align="center"}

#required packages for plot
library(tidyverse)
library(ggplot2)
require(maps)

#Long/Late in same order to nearest site
MLG=c(-5.834874, 56.991962)# check field notes 16
MDC=c(-3.843251, 53.295675) #12
BER=c(-7.213534, 57.713854)#16
TUL=c(-7.025334, 58.185530)# check field notes 16
DGC=c(-4.016982, 57.878096)#14
RVS=c(-2.596053, 56.023971)#6
OBN=c(-5.482266, 56.433569)#14
BWD=c(-3.898247, 50.579755) #devon 7
FRN=c(1.982669, 44.157335) #Will's site in France 6
PCP=c(-4.308992, 51.674219) #Pembrey county park 13
ETB=c( 0.243747, 50.769187) #eastbourne 14
MMS=c(1.255096, 52.168021) #martin's meadows in suffolk 14
CFW=c(-0.281520, 53.254429) #should be Chamber's Farm wood CFW , lincolnshire 10
RHD=c(-1.477433, 54.713907) #Raisby hill durham 13
BMD=c(-1.125439, 51.782563) #Oxford 16


coords <- rbind(MLG, MDC, BER, TUL, DGC, RVS, OBN, BWD, FRN, PCP, ETB, MMS, CFW, RHD, BMD)
colnames(coords) <- c("lon", "lat")
coords <- as.data.frame(coords)
#convert rownames to column
coords <- tibble::rownames_to_column(coords, "POP")

#read in data for samples
samples <- read.table("/media/data_disk/PROJECTS/Saad/CommonBlue/INFO/popmap_sex_copy.tsv", colClasses = c("NULL", "factor", "factor"))
samples <- as.data.frame(table(samples$V2))
colnames(samples) <- c("POP", "Samples")

#combine the data
sampling1 <- inner_join(coords, samples, by="POP")

#filter out data for the FRN site as it won't be plotted
sampling <- sampling1 %>% filter(POP != "FRN")

#Get the region
UK <- map_data("world") %>% filter(region=="UK")

#plot the map
p1 <- ggplot() +
  geom_polygon(data = UK, aes(x=long, y = lat, group = group), fill="grey",alpha=0.3) +
  geom_point( data=sampling, aes(x=lon, y=lat, size=Samples), alpha=0.75) + geom_label(data=sampling, aes(x=lon, y=lat, label=POP), size=3, hjust = 0, nudge_x = 0.3)+
  theme_void() + ylim(50,59) + xlim(NA,2.25) + coord_map() + scale_size_continuous(name="Sample Size", breaks=c(6,10,16))+
  theme(legend.position = c(0.85, 0.8)) 
#print the map
p1
```

**Figure 1** Geographical locations of Great British *Polyommatus icarus* individuals sampled for this study.  An additional 6 individuals collected from southern France are not shown (see Table S2 for full details). Size of circles is proportional to the number of samples acquired at each locality. 

```{r Fig2, warning=FALSE,echo=F, fig.width=9,fig.height=12, message=F, fig.align="centre"}

#additional packages required
library(treeio)
library(ggtree)
library(grid)

x <- read.beast("/media/data_disk/PROJECTS/Saad/CommonBlue/CO1_BOLD/BEAST/modTest_CCS_sc_split_subst/combined.mcc.tree")
setwd("/media/data_disk/PROJECTS/Saad/CommonBlue/scripts/FigScripts")
#rename taxa only to have smaller names
tipnames <- x@phylo$tip.label
boldids <- sapply(strsplit(tipnames, "|", fixed=TRUE),  "[", 1)
country <- sapply(strsplit(tipnames, "|", fixed=TRUE),  "[", 2)
#replace errror and correct formatting
country<- str_replace_all(country, "Polyommatus_icarus", "Italy")
country<- str_replace_all(country, "United_Kingdom", "Great Britain")
d <- data.frame(label = tipnames,
                boldid = boldids, 
                country= country, stringsAsFactors = F)
d$country[d$boldid=="EULEP004-14"] <- "Crete" 

x2 <- full_join(x, d, by = "label")


#for removing bars for with support less than 0.8
x2@data$height_0.95_HPD[x2@data$posterior < 0.75] <- NA

p1 = ggtree(x2) + geom_treescale(x=-1.5, y=100, fontsize=4)  + scale_x_continuous(name="Time (million years)", breaks=c(0,-0.5,-1,-1.5,-2,-2.5,-3),labels = abs, limits=c(NA,0.5)) + theme_tree2(plot.margin=margin(2, 2, 2, 2) )  
p1=revts(p1)
#add uncertainty of time
p2= p1+geom_range('height_0.95_HPD', color='black', size=2, alpha=.25)
#add posterior probaility as number
#p3= p2 +geom_nodelab(aes(x=branch, subset=!isTip & posterior > 0.7, label=round(posterior*100, 0)),hjust=1.5 size=2, nudge_x = 0.08) #+ geom_tiplab(aes(label=country),size=2)
p3 = p2 + geom_text2(aes(label=round(posterior, 2), subset=posterior>0.75), size=3, nudge_x = -0.07, vjust=-.3)

#flip nodes for arrangment
p3<- ggtree::flip(p3,189,216)  
p3<- ggtree::flip(p3,217,317)
p3<- ggtree::rotate(p3, 189) 

#p3 <- p3 + geom_tiplab(aes(label=paste(x2@extraInfo$country)), size=1.9)

#add annoated to clade colours

p6 <- p3 + annotate("text", -2.6, c(180), hjust=0, fontface=2,size=4, label=c(expression(bold("CO1 Lineage"))))+
  annotate("point",-2.5, 170, size=4, shape=21,fill="yellow",color="black") +
  annotate("point",-2.5, 160, size=4, shape=21,fill="green", color="black") +
  annotate("point",-2.5, 150, size=4, shape=21,fill="purple",color="black") +
  annotate("point",-2.5, 140, size=4, shape=21,fill="red",color="black") +
  annotate("text", -2.3,170, hjust=0, size=3.5, label="Iberia-Italy") + 
  annotate("text", -2.3,160, hjust=0, size=3.5,  label= "Palaeartic" )+
  annotate("text", -2.3,150, hjust=0, size=3.5, label="Alicante")+
  annotate("text", -2.3,140, hjust=0, size=3.5, label="Sierra Nevada")

#collapse clades
p6 <- collapse(p6, 195, 'max', fill="Purple", col="black", alpha=0.4) 
p6 <- collapse(p6, 219, 'max', fill="yellow", col="black", alpha=0.4) 
p6 <- collapse(p6, 282, 'max', fill="yellow", col="black", alpha=0.4)
p6 <- collapse(p6, 317, 'max', fill="green", col="black", alpha=0.4) 
p6 <- collapse(p6, 191, 'max', fill="Red", col="black", alpha=0.4)
p6 <- collapse(p6, 299, 'max', fill="white", col="black", alpha=0.4)

#add clade labels with country locations
ibit= c(paste0("Italy"), paste0("Spain"),  paste0("Portugal"), paste0("S. France"))
ibitUK= c(paste0("Great Britain:"), paste0("Scotland (mainland)"),  paste0("Northern England"), paste0("Southern England"), paste0("Finland"))

palaeartic=c(paste0("France, Germany"), paste0("Romania, Finland"), paste0("Norway, Netherlands"), paste0("Iran, Turkey"), paste0("Austria, Canada"),
             paste0("Spain, Italy"), paste0("Greece"))
palaearticUK= c(paste0("Great Britain:"), paste0("Wales"), paste0("Southern England") )

apsnc= c(paste0("Great Britain:"), paste0("Outer Hebrides"), paste0("Scotland (mainland)"), paste0("Alicante, Germany"), paste0("Norway, Kazakhastan"))
unclassified=c(paste0("Italy, Spain"), paste0("Greece, Romania"), paste0("France, Germany"), paste0("Turkey, Iran"))


p7 <- p6  +  annotate("text", x= c(.03,.03,.03,.03) ,y=c(78,75,72,69), hjust=0, size=2.75, label=ibit) +
  annotate("text", x= c(.03,0.07,0.07,.07,.03) ,y=c(35,32,29,26,23), hjust=0, size=2.75, label=ibitUK) +
  annotate("text", x= 0.03 ,y=c(150,147,144,141,138,135,132), hjust=0, size=2.75, label=palaeartic) +
  annotate("text", x= c(.03,0.07,.07) ,y=c(120,117,114), hjust=0, size=2.75, label=palaearticUK) +
  annotate("text", x= 0.03 ,y=187, hjust=0, size=2.75, label="Crete") +
  annotate("text", x= 0.03 ,y=184, hjust=0, size=2.75, label="Sierra Nevada") +
  annotate("text", x= c(.03,.07,.07,.03, 0.03) ,y=c(178,175,172,169,166), hjust=0, size=2.75, label=apsnc) +
  annotate("text", x= c(.03,.03,.03, 0.03) ,y=c(14,11,8,5), hjust=0, size=2.75, label=unclassified)

P8 <- p7 + ggtitle("A") + theme(plot.title = element_text(size= 20, hjust=0.05, vjust=-6, margin = margin(b = 0, t = -1, l = 2, unit = "cm")))

#plotting BOLD sample lineages (based on Dinca et al on map)
bold <- read.csv("../../CO1_BOLD/trimmed_alignments/specimen_info.txt", header=T, na.strings=c("", "NA"), row.names = NULL, sep="\t")
#Keep only UK samples for plotting
UKdata1 <- bold %>% filter(country=="United Kingdom")
#group by lineage and then by lat long
UKdata <- UKdata1 %>% group_by(Lineage.Dinca.et.al..2011., lat,lon) %>%mutate(n= n()) %>% distinct( Lineage.Dinca.et.al..2011.,.keep_all=TRUE)


# Get the world polygon and extract UK
library(maps)
UK <- map_data("world") %>% filter(region=="UK")


p10 <- ggplot() +
  geom_polygon(data = UK, aes(x=long, y = lat, group = group), fill="grey", alpha=0.3) +theme_void() + ylim(50,59)   +
  geom_jitter( data=UKdata, aes(x=lon, y=lat, fill=Lineage.Dinca.et.al..2011., size=n) ,position = position_jitter(width = 0.2, height = 0.2), alpha=0.5, colour="black", pch=21) +
  theme_void() + ylim(50,59) + coord_map() + scale_fill_manual(values=c("purple", "yellow", "green")) + ggtitle("B") + scale_size_continuous(name="No. Specimens", range=c(3,7), breaks=c(1,3,5))+ 
  guides(fill=FALSE) + theme( legend.position = c(0.84, 0.8),legend.title=element_text(size=8),legend.text = element_text(size=7), plot.background = element_rect( colour = "black", size=1),
                              plot.title = element_text(size= 20, hjust=0.05, vjust=-7, margin = margin(b = -0.1, t = -1, l = 2, unit = "cm")) ) 


    
#Plotting map as inset to tree
vp <- viewport(width = .35, height = .35,  x = 0.21, y = 0.22)
print(P8)
print(p10, vp = vp)

```

**Figure 2 (A)** Bayesian ultrametric tree inferred from 187 CO1 *Polyommatus icarus* sequences. Numbers above nodes represent Bayesian posterior support (only where probabilities were > 0.75). The bars on each node represent the 95% Bayesian credible interval for estimates of node age. The scale is in millions of year. Clades with > 0.75 posterior support have been collapsed (see Figure S3 for uncollpased tree) and coloured based on correspondence to the *CO1* lineage classification from Dincă *et al.* (2011). **(B)** Bubble plot of the geographical distribution of the 51 Great British samples in the dataset. Colours are based on their correspondence to the lineage classification from Dincă *et al.* (2011). Size of the bubble represents the total number of CO1 lineages at a particular site.

\pagebreak

```{r Fig3, warning=FALSE, echo=F, fig.width=12,fig.height=12, message=F, results=F, fig.align="centre"}

#load additional packages
library("adegenet")
library(vcfR)
library("ade4")
library(RColorBrewer)
library(ggrepel)
library(dartR)
library(ggpubr)



#VCF <- read.vcfR("/media/data_disk/PROJECTS/Saad/CommonBlue/stacks.denovo/stacks_m4_M4_n4_new/populations.r50.p15_moh_0.65/populations.snps.filter2.0.25.recode.singlesnp.vcf")
#z <- vcfR2genlight(VCF)

#get pops and label 
#popnames <- z@ind.names
#pops <- sapply(strsplit(popnames, '_'), function(x){paste0(x[1])})
#pops<- str_replace_all(pops, "RNL", "CFW")
#pop(z) = as.factor(pops) 
#ploidy(z) <- 2

#add latlong information to genlight object
#latlong = as.data.frame( z@pop)
#latlong$lat <- NULL
#latlong$lon <- NULL

#for (i in 1:dim(latlong)[1]){
#  latlong$lat[i] <- coords[coords$POP==latlong[i,1],2]
#  latlong$lon[i] <- coords[coords$POP==latlong[i,1],3]
  
#}

#names(latlong) <- c("POP","lat", "lon")
#add lat long to gen light object
#z@other$latlong <- latlong[,2:3]
#should now be able to use the dartR IBD function
#add sex information as well
#popnames <- z@ind.names 
#sex<- as.factor(sapply(strsplit(popnames, '_'), function(x){paste0(x[2])}))
#ind <- cbind( z@ind.names, pops, sex, latlong[,2:3])
#colnames(ind)[1] <- "ind"
#z@other$ind.metrics <- ind
#w <- tab(z, freq = TRUE, NA.method = "mean")

#perform PCA
#w.pca <- dudi.pca(w, scannf = F, scale=F, nf=4)

#dudi.pca <- tibble(pc1=w.pca$li[,1], pc2=w.pca$li[,2], pc3=w.pca$li[,3], pc4=w.pca$li[,4],lat=z@other$latlong[,2], pop=pops)
#PCA plot
#pc1per <- round(w.pca$eig[1]/sum(w.pca$eig) *100, 1)
#pc2per <- round(w.pca$eig[2]/sum(w.pca$eig) *100, 1)
#pc3per <- round(w.pca$eig[3]/sum(w.pca$eig) *100, 1)
#pc4per <- round(w.pca$eig[4]/sum(w.pca$eig) *100, 1)
#col <- funky(15)

#labels <- dudi.pca %>% distinct(pop, .keep_all=T)

#current plot with pop colours at random
#p <- ggplot(dudi.pca,aes(x=pc1, y=pc2, colour=pop, group=pop)) + geom_point() +stat_ellipse(level=0.9) + scale_color_manual(name="Pop",values=col) + #coord_fixed(xlim=c(-6, 5), ylim=c(-4.5,5)) +
#  xlab(paste0("PC1 ", pc1per, "%")) + ylab(paste0("PC2 ", pc2per, "%")) + theme_bw() + theme(legend.position = "none",panel.grid = element_blank(), #axis.text=element_text(size=12),axis.title=element_text(size=14,face="bold"), 
                                                                                           #  plot.margin = unit(c(1,1,1,1), "lines"))
#p <- p + geom_label_repel(data=labels, aes(x=pc1, y=pc2, label=pop), size=3)
#add map

#Get the region
#UK <- map_data("world") %>% filter(region=="UK")

#p1 <- ggplot() +
#  geom_polygon(data = UK, aes(x=long, y = lat, group=group), fill="grey",alpha=0.3) +  
#  geom_label(data=coords, aes(x=lon, y=lat, label=POP), size=2.5, hjust = 0,label.size = 0.15 ) + #scale_color_gradient(low = "red", high = "blue", name="Latitude", guide="none")+
#  theme_void() + ylim(50,59) + xlim(NA,3) + coord_map() 

#save the plots
#vp <- viewport(width = .4, height = .4,  x = 0.49, y = 0.72)
#print(p)
#print(p1, vp = vp)


#----------------------------------------------------------------------------
#new figure 3
VCF <- read.vcfR("/media/data_disk/PROJECTS/Saad/CommonBlue/stacks.denovo/stacks_m4_M4_n4_new/populations.r50.p15_moh_0.65/populations.snps.filter2.0.25.recode.vcf")
z <- vcfR2genlight(VCF)
popnames <- z@ind.names
pops <- as.factor(sapply(strsplit(popnames, '_'), function(x){paste0(x[1])}))
pops<- str_replace_all(pops, "RNL", "CFW")
pop(z) = pops
ploidy(z) <- 2
latlong = data.frame( pop=z@pop, lat=numeric(length=length(z@pop)), lon=numeric(length=length(z@pop)))

for (i in 1:dim(latlong)[1]){
  print(i)
  latlong$lat[i] <- coords[coords$POP==latlong[i,1],2]
  latlong$lon[i] <- coords[coords$POP==latlong[i,1],3]
  
}

colnames(latlong) <- c("pop","lat", "lon")
#add lat long to gen light object
z@other$latlong <- latlong[,2:3]
#should now be able to use the dartR IBD function
#add sex information as well
popnames <- z@ind.names 
sex<- as.factor(sapply(strsplit(popnames, '_'), function(x){paste0(x[2])}))
ind <- cbind( z@ind.names, pops, sex, latlong[,2:3])
colnames(ind)[1] <- "ind"
z@other$ind.metrics <- ind
w <- tab(z, freq = TRUE, NA.method = "mean")

#perform PCA
w.pca <- dudi.pca(w, scannf = F, scale=F, nf=4)

dudi.pca <- tibble(pc1=w.pca$li[,1], pc2=w.pca$li[,2], pc3=w.pca$li[,3], pc4=w.pca$li[,4],lat=z@other$latlong[,2], pop=pops)
#PCA plot
pc1per <- round(w.pca$eig[1]/sum(w.pca$eig) *100, 1)
pc2per <- round(w.pca$eig[2]/sum(w.pca$eig) *100, 1)
pc3per <- round(w.pca$eig[3]/sum(w.pca$eig) *100, 1)
pc4per <- round(w.pca$eig[4]/sum(w.pca$eig) *100, 1)
col <- funky(15)

labels <- dudi.pca %>% distinct(pop, .keep_all=T)

#current plot with pop colours at random
p <- ggplot(dudi.pca,aes(x=pc1, y=pc2, colour=pop, group=pop)) + geom_point() +stat_ellipse(level=0.9) + scale_color_manual(name="Pop",values=col) + coord_fixed(xlim=c(-7, 6), ylim=c(-6.5,6)) +
  xlab(paste0("PC1 ", pc1per, "%")) + ylab(paste0("PC2 ", pc2per, "%")) + theme_bw() + theme(legend.position = "none",panel.grid = element_blank(), axis.text=element_text(size=12),axis.title=element_text(size=14,face="bold"), 
                                                                                             plot.margin = unit(c(1,1,1,1), "lines"))
pall <- p + geom_label_repel(data=labels, aes(x=pc1, y=pc2, label=pop), size=3)

library(maps)

UK <- map_data("world") %>% filter(region=="UK")
coords <- as.data.frame(coords)
p1 <- ggplot() +
  geom_polygon(data = UK, aes(x=long, y = lat, group=group), colour="grey",alpha=0.1) +  
  geom_label(data=coords, aes(x=lon, y=lat, label=coords$POP), size=3, hjust = 0,label.size =NA, fill=NA) + #scale_color_gradient(low = "red", high = "blue", name="Latitude", guide="none")+
  theme_void() + ylim(50,59) + xlim(NA,3) + coord_map() 

map_grob <- ggplotGrob(p1)
pall <- pall + annotation_custom(grob=map_grob, xmin=-5.5, xmax=2, ymin=0, ymax=6.5)
#read in Fsts and check outliers
bayesFst <-  read.table("/media/data_disk/PROJECTS/Saad/CommonBlue/stacks.denovo/stacks_m4_M4_n4_new/populations.r50.p15_moh_0.65/BayeScan/p15r50miss25/longrun_n50000_OD100_fst.txt", header = T)
#filter 1% fdr
outliers <- bayesFst[bayesFst$qval<0.01,]
mat <- as.matrix(z)
#use row ids of bayes outliers to select the columnss
outliermat <- mat[,as.numeric(rownames(outliers))]
#append marker information to bayescan outlier for sorting
outliers$marker <- colnames(outliermat)

zneut <- z[,!(z$loc.names %in% outliers$marker)]
w1 <- tab(zneut, freq = TRUE, NA.method = "mean")

#perform PCA
w1.pca <- dudi.pca(w1, scannf = F, scale=F, nf=4)

dudi.pca1 <- tibble(pc1=w1.pca$li[,1], pc2=w1.pca$li[,2], pc3=w1.pca$li[,3], pc4=w1.pca$li[,4],lat=z@other$latlong[,2], pop=pops)
#PCA plot
pc1per <- round(w1.pca$eig[1]/sum(w1.pca$eig) *100, 1)
pc2per <- round(w1.pca$eig[2]/sum(w1.pca$eig) *100, 1)
pc3per <- round(w1.pca$eig[3]/sum(w1.pca$eig) *100, 1)
pc4per <- round(w1.pca$eig[4]/sum(w1.pca$eig) *100, 1)
col <- funky(15)

labels <- dudi.pca1 %>% distinct(pop, .keep_all=T)

#current plot with pop colours at random
p <- ggplot(dudi.pca1,aes(x=pc1, y=pc2, colour=pop, group=pop)) + geom_point() +stat_ellipse(level=0.9) + scale_color_manual(name="Pop",values=col) + coord_fixed(xlim=c(-7, 6), ylim=c(-6.5,6)) +
  xlab(paste0("PC1 ", pc1per, "%")) + ylab(paste0("PC2 ", pc2per, "%")) + theme_bw() + theme(legend.position = "none",panel.grid = element_blank(), axis.text=element_text(size=12),axis.title=element_text(size=14,face="bold"), 
                                                                                             plot.margin = unit(c(1,1,1,1), "lines"))
pneut <- p + geom_label_repel(data=labels, aes(x=pc1, y=pc2, label=pop), size=3)
#pca with outlier loci
zout <- z[,outliers$marker]
w2 <- tab(zout, freq = TRUE, NA.method = "mean")

#perform PCA
w2.pca <- dudi.pca(w2, scannf = F, scale=F, nf=4)

dudi.pca2 <- tibble(pc1=w2.pca$li[,1], pc2=w2.pca$li[,2], pc3=w2.pca$li[,3], pc4=w2.pca$li[,4],lat=z@other$latlong[,2], pop=pops)
#PCA plot
pc1per <- round(w2.pca$eig[1]/sum(w2.pca$eig) *100, 1)
pc2per <- round(w2.pca$eig[2]/sum(w2.pca$eig) *100, 1)
pc3per <- round(w2.pca$eig[3]/sum(w2.pca$eig) *100, 1)
pc4per <- round(w2.pca$eig[4]/sum(w2.pca$eig) *100, 1)
col <- funky(15)

labels <- dudi.pca2 %>% distinct(pop, .keep_all=T)

#current plot with pop colours at random
p <- ggplot(dudi.pca2,aes(x=pc1, y=pc2, colour=pop, group=pop)) + geom_point() +stat_ellipse(level=0.9) + scale_color_manual(name="Pop",values=col) + coord_fixed(xlim=c(-3, 3), ylim=c(-3,3)) +
  xlab(paste0("PC1 ", pc1per, "%")) + ylab(paste0("PC2 ", pc2per, "%")) + theme_bw() + theme(legend.position = "none",panel.grid = element_blank(), axis.text=element_text(size=12),axis.title=element_text(size=14,face="bold"), 
                                                                                             plot.margin = unit(c(1,1,1,1), "lines"))
pout <- p + geom_label_repel(data=labels, aes(x=pc1, y=pc2, label=pop), size=3)


#plot frequencies of most common allele
#drawing allele frequencies across lats, pops
snps <- data.frame(POP=pop(z))
snps <- cbind(snps, outliermat)

#reorder pops in lat order
snps$POP <- factor(snps$POP, c("FRN", "ETB", "BWD", "BMD", "PCP", "MMS", "CFW", "MDC", "RHD", "RVS", "OBN", "MLG", "DGC", "BER", "TUL"))

allelefreqs <- data.frame(Pops=levels((snps$POP)))
allelefreqs$Pops <- factor(allelefreqs$Pops, c("FRN", "ETB", "BWD", "BMD", "PCP", "MMS", "CFW", "MDC", "RHD", "RVS", "OBN", "MLG",  "BER",  "DGC","TUL"))
for (i in 2:dim(snps)[2]){
  #get table of allele frequencies

  x<-xtabs(~snps[,i]+snps$POP)
  #pick the most common allele
  #if there are only two classeses
  if (dim(x)[1]==3) { 
    #most common allele
    if (sum(x[1,])>sum(x[3,])){
      mca_ind = 1
      other_ind =3
    }
    else {mca_ind=3; other_ind=1 }
    }
  
  
  if (dim(x)[1]==2){
    print( colnames(snps)[i])
    freq <- (2*x[1,]+x[2,])/(2*(x[1,]+x[2,]))
    allelefreqs$marker <- freq
    names(allelefreqs)[i] <- names(snps)[i]
  }
  else {
  #print( colnames(outliermat)[i])
  freq <- (2*x[mca_ind,]+x[2,])/(2*(x[mca_ind,]+x[2,]+x[other_ind,]))
  allelefreqs$marker <- freq
  names(allelefreqs)[i] <- names(snps)[i]}
}


#plot all the outliers
library(reshape2)
d <- melt(allelefreqs, id.vars = "Pops")
#filter France and other small pops
d <- d %>% filter(Pops!="FRN" & Pops!="RVS" & Pops!="BWD")
#clines
#ggplot(d %>% filter(variable=="366257_57"), aes(x=Pops, y=value, group=variable)) + geom_line(alpha=0.2)
#merge pops
d$Region <- "South"
d$Region[d$Pops=="BER" | d$Pops=="TUL" | d$Pops=="DGC" | d$Pops=="MLG" | d$Pops=="OBN"] <- "North"
d1 <- d %>% group_by(Region, variable) %>% summarise(mean=mean(value, na.rm=T)) 
mca <- ggplot(d1, aes(x=mean, fill=Region)) + geom_density(alpha=0.6) + theme_bw()+  xlab("Frequency of Most Common Allele") + 
  scale_fill_manual(values=c("black", "grey"))+
    theme(legend.position =c(0.7,.7),panel.grid = element_blank(), 
          axis.text=element_text(size=12),axis.title=element_text(size=13,face="bold"), 
          legend.title = element_text( size = 14),
          legend.text = element_text( size = 12),
          legend.spacing.y = unit(-0.25, "mm"),
          legend.key.size = unit(1,"line"),
          plot.margin = unit(c(1,1,1,1), "lines"))

ggarrange(pall,pneut, pout, mca, labels = c("A", "B", "C", "D"), ncol=2, nrow=2,font.label = list(size = 20, color = "black", face = "bold", family = NULL))


```

**Figure 3** Population structure based on principal component analysis (PCA) of 5592 SNPs (2824 RAD loci) of 148 *Polyomattus icarus* individuals across the British Isles. PCAs are shown for the entire dataset **(A)**, a subset of 5479 putatively neutral SNPS **(B)**, and a subset of 113 outlier loci **(C)**. Ninety-% Confidence Interval (CI) ellipses for PC1 and PC2 for each locality are also shown. Each locality is represented by a distinct colour and map inset in **(A)** shows a labelled map of the sampled localities for quick reference. **(D)** Density distributions of the frequency of the most common allele (mca) for the 113 outlier loci across northern and southern British Isles. Indvidual Sampling locations are aggregated by North-South division in **(C)** and localities with less than 7 individuals were excluded (BWD, RVS, FRN). North: TUL, BER, DGC, MLG, OBN; South: MDC, CFW, MMS, PCP, BMD, ETB.


```{r Fig4, warning=FALSE, echo=F, fig.width=5,fig.height=8, message=F, results=F, fig.align="center"}
#load additional packages required
library(gridExtra)


setwd("/media/data_disk/PROJECTS/Saad/CommonBlue/scripts/FigScripts")

wolbachia <- read.csv("../../Wolbachia/Wolbachia_infection.csv", sep="\t")

#create a binary variable for whether individual was infected or not based on threshold of log(percent)>1

wol <- wolbachia %>% mutate(infected=ifelse(log(percentage_classified_wolcbachia)>1,1,0))

#create column for male/females
wol <- wol %>% mutate(Sex=str_split(X.sample,"_", simplify = T)[,2])

#replace RNL with CFW
wol$POP<- str_replace_all(wol$POP, "RNL", "CFW")


#suummarize infection levels
infection <- as.data.frame.matrix(table(wol$POP, wol$infected))
names(infection) <- c("un", "inf")
infection$prop <- infection$inf/(infection$un+infection$inf)
infection$POP <- rownames(infection)
#merge the lat long info for plotting
infection <- full_join(infection, coords, by = "POP")
#calculate standard errors for the proportion
infection <- infection %>% mutate(se=sqrt((prop*(1-prop))/(inf+un)))
#note france won't be drawn
UK <- map_data("world") %>% filter(region=="UK")

#plot the map
p1 <- ggplot() +
  geom_polygon(data = UK, aes(x=long, y = lat, group = group), fill="grey",alpha=0.3) +
  geom_point( data=infection%>%filter(prop>0), aes(x=lon, y=lat, size=prop),col="red", alpha=0.7) + geom_label(data=infection, aes(x=lon, y=lat, label=POP), size=3, hjust = 0, nudge_x = 0.35)+
  theme_void() + ylim(50,59) + xlim(NA,2.2) + coord_map() +scale_size_continuous(name="Proportion of Individuals Infected", breaks=c(0.25,.50,.75), range=c(5,10)) + 
  theme(legend.position = c(0.85, 0.8))
#p1


#barplot with map inset
#relevel the order for POP
infection$POP <- factor(infection$POP, c("RHD", "RVS", "OBN", "MLG", "DGC", "BER", "TUL"))
ibp<-ggplot(infection %>% filter(prop>0), aes(x=POP, y=prop)) + geom_bar(stat="identity", alpha=0.5) + theme_bw() +
ylim(c(0,1)) + xlab("Population") + ylab("Proportion of Individuals Infected")+ coord_flip() +
geom_errorbar(aes(x=POP, ymin=prop-se, ymax=prop+se), width=0.1, colour="black", alpha=0.5, size=0.75)+
theme(axis.text=element_text(size=10),axis.title=element_text(size=12,face="bold"), 
      plot.margin = unit(c(1,1,1,1), "lines"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank())



#plot the map
library(sf)
library("rnaturalearth")
library("rnaturalearthdata")
require(rgeos)
library("ggspatial")

UK2 <- ne_countries(scale = "medium", returnclass = "sf", country="united kingdom")


p3 <- ggplot(data = UK2) +
  geom_sf(alpha=0.1, colour="gray") + theme_void() +  coord_sf( xlim=c(-8,.6),ylim = c(54.5, 59), expand = TRUE) +
  geom_label(data=infection, aes(x=lon, y=lat, label=POP), size=2.2, hjust = 0, label.size = NA, fill=NA, face="bold") +
  theme(plot.background = element_rect( colour = "black", size=1))

library(grid)
ibpm <- ibp+
  annotation_custom(ggplotGrob(p3), xmin = -.5, xmax = 4, 
                    ymin = 0.7, ymax = 1.03)
# ggtitle("A") plot.title = element_text(size= 20, hjust=0.05, vjust=-7, margin = margin(b = -0.1, t = -1, l = 2, unit = "cm"))
#boxplot for infection levels of norther populations

north <- wol %>% filter(POP=="BER" | POP=="TUL" | POP== "DGC" | POP=="MLG" | POP=="OBN")


#summarize the data
summ_north <- north%>% group_by(POP, Sex) %>% 
  mutate(mean=mean(percentage_classified_wolcbachia), stde=sd(percentage_classified_wolcbachia)/sqrt(n())) %>%
  distinct(POP,.keep_all=TRUE)

#relevel the order for POP
north$POP <- factor(north$POP, c("OBN", "MLG", "DGC", "BER", "TUL"))


p2 <- ggplot(north, aes(y=percentage_classified_wolcbachia, x=POP, group=Sex, colour=Sex)) +
      geom_jitter(size=2, alpha=0.5, width = 0.25) +  coord_flip() + theme_bw() + scale_colour_manual(values=c("black", "grey80"))+
      xlab("Population") + ylab("Percentage of Classified Reads Mapped to Wolbachia") + geom_point(data=summ_north, aes(y=mean, x=POP, fill=Sex),colour="black",size=5, pch=21, alpha=0.5,position=position_dodge(width=0.5))+
      #geom_errorbar(data=summ_north,aes(ymin = mean - stde, ymax = mean + stde),width = 0.25, size = 0.5,position=position_dodge(width=0.5) )+
      guides(colour=FALSE) + scale_y_continuous(breaks=c(0,10,20,30,40,50))+scale_fill_manual(values=c("black", "grey80"))+
      theme(legend.position=c(0.85,0.25),axis.text=element_text(size=10),axis.title=element_text(size=12,face="bold"), 
            plot.margin = unit(c(1,1,1,1), "lines"),
            legend.title = element_text( size = 11, hjust = 0.25),
            #legend.text = element_text( size = 9),
            legend.spacing.y = unit(-0.5, "mm"),
            legend.key.size = unit(2,"line"),
            legend.text = element_text(margin = margin(r = 8, unit = "pt")),
            #legend.box.background = element_rect(colour = "black"),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank())


#print the plots
ggarrange(ibpm,p2, labels = c("A", "B"), nrow=2,font.label = list(size = 15, color = "black", face = "bold", family = NULL))

```


**Figure 4 (A)** Proportion of individuals classified as infected based on a threshold of  log2 (percentage of classified reads mapping to *Wolbachia*) > 0. Only populations with at least one infected individual are shown. Error bars are the standard errors of the estimated proportions. Inset shows the geographical range of populations with infected individuals. **(B)** Distributions of the percentage of classified reads mapping to *Wolbachia* for locations with individuals classified as infected. RHD with a single infected male is not shown. Larger circles represent the average for each sex within locations.

\pagebreak

```{r Fig5, warning=FALSE,echo=F, fig.width=10,fig.height=10, message=F, fig.align="center", results=F}

#no additional packages required
#read in the saved trees in newick format
wol_tree <- read.newick("woltree_more.nwk", node.label = "support")
co1_tree <- read.newick("co1tree_more.nwk", node.label = "support")
p1 <- ggtree(wol_tree)
p2 <- ggtree(co1_tree)

#add popinformation
#get pop names
#get pop names
pops <- sapply(strsplit(wol_tree@phylo$tip.label, '_'), function(x){paste0(x[1])})
wol <- as_tibble(wol_tree)
dat <- tibble(label=as.character(wol_tree@phylo$tip.label), pop=pops) %>% mutate(geog=ifelse(pop=="BER" | pop =="TUL", "OH", "Main"))
y <- full_join(wol, dat, by = 'label')
wol_tree2 <- as.treedata(y)

#see node labels

#p1  + geom_label2(aes(subset=!isTip, label=node), size=3, color="darkred", alpha=0.5) 

p1 <- ggtree(wol_tree2) + geom_text2(aes(subset =  (support > 70), label=support), hjust=-0.1)
p1 <- p1 + geom_tippoint(aes(colour=pop, group=pop,size=geog),fill="gray80", shape=19, show.legend =TRUE ) + scale_size_manual(values=c(3.5, 1.5)) +
  scale_colour_manual(values=c("#A6CEE3", "#4F9F3B","#FDAC4F", "#D1AAB7", "#A99099","#B15928"  ))
#p1 + geom_treescale()

#construct the CO1 tree
pops <- sapply(strsplit(co1_tree@phylo$tip.label, '_'), function(x){paste0(x[1])})
co1<- as_tibble(co1_tree)
dat2 <- tibble(label=as.character(co1_tree@phylo$tip.label), pop=pops) %>% mutate(geog=ifelse(pop=="BER" | pop =="TUL", "OH", "Main"))
y <- full_join(co1, dat2, by = 'label')
co1_tree2 <- as.treedata(y)

#see node labels

#p2  + geom_label2(aes(subset=!isTip, label=node), size=3, color="darkred", alpha=0.5) 
#make the tree
p2 <- ggtree(co1_tree2)  + geom_text2(aes(subset =  (support > 97) & !(isTip), label=support), hjust=1.3,  vjust=-.5)
p2 <- p2 + geom_tippoint(aes(colour=pop, group=pop,size=geog),fill="gray80", shape=19, show.legend =TRUE ) + scale_size_manual(values=c(3.5, 1.5)) +
  scale_colour_manual(values=c("#A6CEE3", "#4F9F3B","#FDAC4F", "#D1AAB7", "#A99099","#B15928"  )) 

#geom_hilight_encircle(node=33, fill='yellow',alpha=0.6) +
#geom_hilight_encircle(node=46, fill='purple', alpha=0.6) 
#p2 =revts(p2)

d1 <- p2$data
d2 <- p1$data

## reverse x-axis and 
## set offset to make the tree in the right hand side of the first tree
d2$x <- max(d2$x) - d2$x + max(d1$x) + 1
#manually adjust tree positions
d2$x <-d2$x - 0.9


pp <- p2  + geom_tree(data=d2)   + geom_text2(aes(subset =  (support > 70) & !(isTip), label=support), data=d2, hjust=-.3,  vjust=-.5)+
  geom_tippoint(aes(size=geog, colour=pop, group=pop),data=d2,fill="gray80", shape=19, show.legend =TRUE )  #+
#geom_hilight_encircle(node=33, data=d2,fill='yellow',alpha=0.6) +
#geom_hilight_encircle(node=46, data=d2,fill='purple', alpha=0.6) 

dd <- bind_rows(d1, d2) %>% 
  filter(!is.na(label))

pp <- pp + geom_line(aes(x, y, group=label), data=dd, color='grey', alpha=0.5)

#save tree image
#pdf(file = "FigS8.pdf", width = 9, height = 11)
#annoatate tree
pp + annotate("text", 0.34,36, hjust=0, size=6, label=expression(atop(italic("Wolbachia ") * "ddRAD Loci", "(37 loci, 74 SNPs)"))) +
  annotate("text", 0.0,19, hjust=0, size=6, label=expression(atop("Mitochondrial CO1", "(16 SNPs)"))) +
  geom_cladelabel(node=46, label="",offset=0.01,align=TRUE, colour="Purple" ,hjust=-0.1, barsize = 2, alpha=0.5)+ 
  geom_cladelabel(node=33,label="",offset=0.01,align=TRUE, colour="Yellow", angle=90 ,hjust=-0.1, barsize = 2, alpha=0.5)+ 
  #geom_cladelabel(node=46, label="Alicante-Sierra Nevada", offset=0.02,align=TRUE, color='black', angle=270, offset.text = 0.01)+ 
  #theme(legend.position = c(0.1,0.85))
  annotate("text", x= -0.01 ,y=55, hjust=0, size=6, label="Population", Parse=TRUE) +
  annotate("point", x= 0.0 ,y=52, hjust=0, size=2, shape=19, colour="#A6CEE3", Parse=TRUE) +
  annotate("text", x= 0.02 ,y=52, hjust=0, size=4, label="BER (Outer Hebrides)", Parse=TRUE) +
  annotate("point", x= 0.0 ,y=49, hjust=0, size=2, shape=19, colour="#B15928", Parse=TRUE) +
  annotate("text", x= 0.02 ,y=49, hjust=0, size=4, label="TUL (Outer Hebrides)", Parse=TRUE) +
  annotate("point", x= 0.0 ,y=46, hjust=0, size=4, shape=19, colour="#4F9F3B", Parse=TRUE) +
  annotate("text", x= 0.02 ,y=46, hjust=0, size=4, label="DGC (Northern Scotland mainland)", Parse=TRUE) +
  annotate("point", x= 0.0 ,y=43, hjust=0, size=4, shape=19, colour="#FDAC4F", Parse=TRUE) +
  annotate("text", x= 0.02 ,y=43, hjust=0, size=4, label="MLG (Northern Scotland mainland)", Parse=TRUE) +
  annotate("point", x= 0.0 ,y=40, hjust=0, size=4, shape=19, colour="#D1AAB7", Parse=TRUE) +
  annotate("text", x= 0.02 ,y=40, hjust=0, size=4, label="OBN (Northern Scotland mainland)", Parse=TRUE) +
  annotate("point", x= 0.0 ,y=37, hjust=0, size=4, shape=19, colour="#A99099", Parse=TRUE) +
  annotate("text", x= 0.02 ,y=37, hjust=0, size=4, label="RHD (Northern England)", Parse=TRUE) + 
  annotate("text", x=0.23, y=8, angle=270, size =4, label="Iberia-Italy")+
  annotate("text", x=0.23, y=23, angle=270, size =4, label="Alicante-Provence")+
  annotate("segment", x = 0.29, xend = 0.29, y = 29, yend = 55, colour="black", size=2, alpha=0.5) +
  annotate("segment", x = 0.29, xend = 0.29, y = 1, yend = 27, colour="black", size=2, alpha=0.5) +
  annotate("text", x=0.28, y=41, angle=90, size =4, label=expression(italic("w")*"Ica1"))+
  annotate("text", x=0.28, y=14, angle=90, size =4, label=expression(italic("w")*"Ica2"))+ 
  geom_treescale(x=0.4, y=42, fontsize=5) 
#dev.off()
```


**Figure 5** UPGMA clustering of mitochondrial CO1 fragments from 31 *Polyommatus icarus* individuals based on bitwise distance (left). UPGMA clustering of 74 concatenated SNPs from ddRADseq *Wolbachia* genotypes derived from reads mapping to Wolbachia from 55 infected individuals (right). Numbers on nodes represent bootstrap branch support values based on 1000 bootstrap replicates, values < 70% are not shown. Large circles represent individuals from mainland populations (DGC, MLG, OBN, RHD) and smaller circles represent individuals from the Outer Hebrides (BER, TUL). Lines between dendograms connect the CO1 haplotype and *Wolbachia* straind for the same individual. Scale bar reflects the proportion of loci that are different.

\pagebreak

```{r Fig6, warning=FALSE,echo=F, fig.width=10,fig.height=4, message=F, fig.align="center"}

#read the marker data
sexm <- read.csv("sexmarkerdata.csv")

sexm$geno.sex <- ifelse(sexm$geno.sex=="f", "Phenotypic Female", "Phenotypic Male" )
names(sexm)[4] <- "Genotype"
names(sexm)[3] <- "Counts"
names(sexm)[1] <- "infection"
#filter redundnat markers
#sexm <- sexm %>% filter(Marker!="24861_43" & Marker != "24861_17" & Marker != "22073_63")

#exemplify one marker
sexm %>% filter(markers=="11011_27") %>% ggplot(aes(fill=Genotype, y=Counts, x=infection, group=geno.sex)) + geom_bar(stat="identity", alpha=0.5) +
          theme_bw() + scale_fill_manual(values=c("black", "grey80"))+ labs(fill="Genotypic Class:")+
          #scale_x_discrete(labels=c("Uninfected", expression(italics("w")*"Ica1"), expression(italics("w")*"Ica2"))) + 
          xlab("Infection Status") +
          theme(legend.position="top",panel.grid.major = element_blank(),panel.grid.minor = element_blank(), legend.text = element_text( size = 12),
                strip.background = element_rect(colour="black", fill="white", size=1.5, linetype = "solid"), legend.title = element_text( size = 14),
                strip.text = element_text(size=12, face="bold"),
                axis.text=element_text(size=12, face="bold"),axis.title=element_text(size=14,face="bold"))+
          facet_wrap(~geno.sex, ncol=2, scales="free")

```

**Figure 6** Feminization of *w*Ica1 infected females as suggested by discordance between phenotypic sex, based on external genitalia and dimorphic wing patterning, and genetic sex based on female-specific markers. Discordance is exemplified by data on a single marker here (11011_27), see Table S7 for more examples. All *w*Ica1 infected morphological females (all females in the Outer Hebrides and one from the mainland(MLGf010)) are homozygous for female-specific markers, consistent with feminization.